Studio: handle errors when sending messages

katinthehatsite opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Closes: https://github.com/Automattic/dotcom-forge/issues/7668

## Proposed Changes

This PR adds UI to the AI assistant for handling errors when there is a failure in the API request:

<img width="998" alt="Screenshot 2024-06-25 at 2 30 53 PM" src="https://github.com/Automattic/studio/assets/25575134/bb4e0dae-5512-4b6e-b2ba-6d0943c3a53b">

## Testing Instructions

* Pull the changes from this branch locally
* Modify the URL for the API to be invalid in `src/hooks/use-assistant-api.ts` on lines 50-54:

```
						{
							path: '/studio-app/ai-assistant/cht',
							apiNamespace: 'wpcom/v2',
							body,
						},
```
* Navigate to the Assistant tab
* Try sending either a message or use an example prompt
* Observe that error message `Oops! We couldn't get a response from the assistant. Try again` appears
* Click on `Try again` link
* Observe that it attempts to send a message again and the thinking mode is on
* Remove the changes from `src/hooks/use-assistant-api.ts` and confirm that sending a message works as expected and nothing is broken in the usual assistant flow

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

Commit:
Add Error Notice and remove default window

Commit:
Add link and handle try again

Commit:
Adjust assistant thinking and errors

Commit:
Fix alignment of try again

Commit:
Fix font sizing on error message

Commit:
Apply error message styling

Commit:
Add colors

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -173,35 +214,39 @@ export function ContentTabAssistant( { selectedSite }: ContentTabAssistantProps
 		if ( chatMessage.trim() ) {
-			addMessage( chatMessage, 'user', chatId );
+			const newMessageId = messageId ?? messages.length;
+			if ( messageId === undefined ) {
```
I usually use the following approach when checking if a variable is undefined as it covers more scenarios:

```suggestion
			if ( typeof messageId === 'undefined' ) {
```

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -173,35 +214,39 @@ export function ContentTabAssistant( { selectedSite }: ContentTabAssistantProps
+				addMessage( chatMessage, 'user', chatId );
+			}
 			setInput( '' );
+			setIsThinking( true );
```
I'm not sure if we need to keep this in the component state. In theory, `isAssistantThinking` should represent the state of waiting for the response. If it fails, it will set it to `false` ([reference](https://github.com/Automattic/studio/blob/880ff2d8301924eafdc19ecee40fc1fbf8b0b805/src/hooks/use-assistant-api.ts#L70)).

katinthehatsite replied:
The main reason why I had to add it is to deal with re-trying sending the message. I have now refactored my approach quite a bit so I will check if it can be removed

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -94,17 +126,26 @@ const AuthenticatedView = memo(
+						>
+							{ message.content }
+						</ChatMessage>
+						{ message.id === errorMessageId && (
+							<ErrorNotice
+								handleSend={ handleSend }
+								messageContent={ message.content }
+								messageId={ message.id }
+							/>
+						) }
```
I presume that when an error is encountered, it will always be associated with the latest message. If so, we don't really need to check the message ID, we could just render the error notice after the last message.

sejas replied:
We should support the scenario where the last message fails, the user sends another message, and it succeeds, so it's not guaranteed that the failing message will always be the last message.

I'm okay if we show the "try again" message only if it is the last one, if that makes things easier. However, the chat history should display which messages failed and were not sent.

fluiddot replied:
@sejas But if we send all messages on every request ([reference](https://github.com/Automattic/studio/blob/a59527b5c5a997332bcbd061a92b56dbe71da33f/src/hooks/use-assistant-api.ts#L31-L38)), can't we consider any previous failed message as succeeded if the last one succeeded? I might be overlooking how the system works, so please correct me if my assumption is not accurate.

fluiddot replied:
Following [Matt's comment](https://github.com/Automattic/studio/pull/297#issuecomment-2194039268), it would be great if specify what to do with failed messages after receiving successful response. Here are the options I identified:

1. Failed messages are kept in the chat history and marked as failed. The retry message notice won't show up.
2. Failed messages are removed.
3. Failed messages are kept in the chat history but marked as successful.

For the first and second options, we'd need to tweak the messages we sent to the AI as currently we send all, including the failed.

cc @matt-west @katinthehatsite

matt-west replied:
> Failed messages are kept in the chat history but marked as successful

This is my preference, but I didn't communicate it as effectively in my comment.

If I send a message and it fails, I should see the error state and a `Try again` button. Once I click the `Try again` button, the error state is immediately removed while we wait for a response from the assistant. If it fails again, the error state is added back to the same message. If it succeeds, we continue as normal.

The key is that there's only ever one message shown in the chat history, regardless on how many times it has failed and been retried.

sejas replied:
> The key is that there's only ever one message shown in the chat history, regardless on how many times it has failed and been retried.

It makes sense to keep only one failed message if a retry fails.
But the user can write in the text input and send a new message that will fail. In that case the user will see two failed messages.
To simplify the logic we'll display the retry button to the failed message if it's the last one. So only a retry button that will re-send the latest message.

We discussed to disable the text input if there is a failed message. But I think that's extreme even though that's how ChatGPT works:

<img width="1709" alt="Screenshot 2024-06-27 at 14 20 23" src="https://github.com/Automattic/studio/assets/779993/6738d03d-2be7-4684-b9d0-541e256ac1a0">

katinthehatsite replied:
> The key is that there's only ever one message shown in the chat history, regardless on how many times it has failed and been retried.

Nice, thanks for clarifying this. I think that's the part that was confusing me as I was not sure if it was stay just as one message or should be resent multiple times. I will go for this implementation.

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -173,35 +214,39 @@ export function ContentTabAssistant( { selectedSite }: ContentTabAssistantProps
 	const { __ } = useI18n();
 	const lastMessage = messages.length === 0 ? undefined : messages[ messages.length - 1 ];
 
+	const [ errorMessageId, setErrorMessageId ] = useState< number | null >( null );
```
As shared above, we could consider simply keeping a boolean in the component state. We could have something like:
```suggestion
	const [ hasError, setHasError ] = useState( false );
```

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -173,35 +214,39 @@ export function ContentTabAssistant( { selectedSite }: ContentTabAssistantProps
 	}, [ fetchWelcomeMessages, selectedSite ] );
 
-	const handleSend = async ( messageToSend?: string ) => {
+	const handleSend = async ( messageToSend?: string, messageId?: number ) => {
```
I wouldn't expect the `messageId` argument to be part of handling sending a message, as the id is set by the AI. That said, I wonder if we could handle the message retry in a separate function to simplify it.

E.g.:
```
	const onRetryLastMessage = async () => {
		setHasError( false );
		try {
			const { message, chatId: fetchedChatId } = await fetchAssistant(
				chatId,
				messages,
				currentSiteChatContext
			);
			if ( message ) {
				addMessage( message, 'assistant', chatId ?? fetchedChatId );
			}
		} catch ( error ) {
			setHasError( true );
		}
	};
```

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -173,35 +214,39 @@ export function ContentTabAssistant( { selectedSite }: ContentTabAssistantProps
-					[ ...messages, { content: chatMessage, role: 'user', createdAt: Date.now() } ],
+					[
+						...messages,
+						{ content: chatMessage, role: 'user', createdAt: Date.now(), id: newMessageId },
```
When retrying a message we'll append the same message to the AI. This might have side effects on the response, so I'd propose just sending the same messages before the error.

fluiddot reviewed (commented):
I also noticed that the error state doesn't reflect the designs, specifically:
- The message doesn't have the red styling.
- The Try again button is too close to the text.

**App:**
<img width="421" alt="Screenshot 2024-06-25 at 18 42 44" src="https://github.com/Automattic/studio/assets/14905380/cc1d6083-beda-4fa6-bcc3-ce76ff530868">

**Design:**
<img width="737" alt="Screenshot 2024-06-25 at 18 45 54" src="https://github.com/Automattic/studio/assets/14905380/2bff0652-edcc-4492-be5e-480e2ec95fb2">

katinthehatsite commented:
@fluiddot - thanks for the review! Let me rethink the approach here and I will re-request the review after. There are some edge cases that seems to need to be covered.

fluiddot commented:
> @fluiddot - thanks for the review! Let me rethink the approach here and I will re-request the review after. There are some edge cases that seems to need to be covered.

Great, let me know if I can help out anyhow in the potential changes.

Commit:
Add errors to every message that fails to send

Commit:
Fix styling on error messages

Commit:
Pass hasError prop to ChatMessage

Commit:
Persist error notices

Commit:
Remove is thinking state

katinthehatsite commented:
Hey @fluiddot and @sejas - I have changed my approach a bit based on your comments:

* I have now added the notices to each message that fails to send so if multiple messages fail to send, the notice will appear under each one of them with `Try again` button
* It should persist by the messages that fail when you navigate away from the tab and come back as I am saving the error notices to `localStorage`
* Clicking the `Try again` button will re-send the message
* One thing that I have not done/not sure about if the failed message should disappear if it was successfully sent

Feel free to try it out and let me know what you think of this approach!

matt-west commented:
Thanks @katinthehatsite! This looks great.

> One thing that I have not done/not sure about if the failed message should disappear if it was successfully sent

I think we should remove the failed message if it is successfully resent. Otherwise you could end up with a chat history filled with the same message multiple times.

<img width="1012" alt="Screenshot 2024-06-27 at 08 56 59" src="https://github.com/Automattic/studio/assets/417538/504f949a-caf5-4f61-a7aa-66369f564165">

katinthehatsite commented on src/components/content-tab-assistant.tsx:
```diff
@@ -21,14 +21,24 @@ import { ChatMessage } from './chat-message';
 const UsageLimitReached = () => {
 	const { daysUntilReset } = usePromptUsage();
 
-	// Determine if the reset is today
```
I noticed that I accidentally removed this comment, will add it back

fluiddot commented:
> > One thing that I have not done/not sure about if the failed message should disappear if it was successfully sent
> 
> I think we should remove the failed message if it is successfully resent. Otherwise you could end up with a chat history filled with the same message multiple times.
> 
> <img alt="Screenshot 2024-06-27 at 08 56 59" width="1012" src="https://private-user-images.githubusercontent.com/417538/343682763-504f949a-caf5-4f61-a7aa-66369f564165.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0NzU0NzQsIm5iZiI6MTcxOTQ3NTE3NCwicGF0aCI6Ii80MTc1MzgvMzQzNjgyNzYzLTUwNGY5NDlhLWNhZjUtNGY2MS1hN2FhLTY2MzY5ZjU2NDE2NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyN1QwNzU5MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNjAzOTRiZmI0ZWFjMGJmMzliMDVjMGIyMjE0NDFiM2Q4MjJhNzVmZmYwYmQ0MWI2OWMyNGIzODIwYmU3MjUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vNPN9iI2Ks2tyIxqL1ibjswRY_FwPFJ6i58_gTjAJEs">

For reference, in [this comment thread](https://github.com/Automattic/studio/pull/297#discussion_r1654611580) @sejas and I discussed about this topic.

katinthehatsite commented:
> I think we should remove the failed message if it is successfully resent. Otherwise you could end up with a chat history filled with the same message multiple times.

I am attempting to implement this atm. I think this would be a nice behavior which would be similar to many messaging platforms. I will update the PR once it works the way I am expecting it to.

fluiddot commented on src/components/chat-message.tsx:
```diff
@@ -158,9 +160,12 @@ export const ChatMessage = ( {
+					'inline-block p-3 rounded overflow-x-auto select-text',
 					isUnauthenticated ? 'lg:max-w-[90%]' : 'lg:max-w-[70%]', // Apply different max-width for unauthenticated view
-					! isUser ? 'bg-white' : 'bg-white/45'
+					! isUser ? 'bg-white' : '',
+					isUser && ! hasError ? 'bg-white/45' : '',
```
I think that for these two statements, it's not necessary to use the ternary operator. We can do:
```suggestion
					! isUser && 'bg-white',
					isUser && ! hasError && 'bg-white/45',
```

fluiddot commented on src/components/chat-message.tsx:
```diff
@@ -158,9 +160,12 @@ export const ChatMessage = ( {
+					! isUser ? 'bg-white' : '',
+					isUser && ! hasError ? 'bg-white/45' : '',
+					hasError ? 'border-[#FACFD2] bg-[#F7EBEC]' : 'border-gray-300',
+					'border'
```
Do we need to place this class name at the end of the argument list? If not, probably we could move it to the first item.

fluiddot commented on src/components/content-tab-assistant.tsx:
```diff
@@ -94,17 +138,26 @@ const AuthenticatedView = memo(
+						>
+							{ message.content }
+						</ChatMessage>
+						{ message.id !== undefined && (
+							<ErrorNotice
+								handleSend={ handleSend }
+								messageContent={ message.content }
+								hasError={ errorNotices.has( message.id ) }
+							/>
+						) }
```
As shared [here](https://github.com/Automattic/studio/pull/297#issuecomment-2194039268), to avoid displaying this notice multiple times, we could move the component outside the map.

For reference, I'm working on a refactor of this component that aims to unify notice rendering (https://github.com/Automattic/studio/blob/13e489e812298dc07a16a7584a6547aa2bc6157d/src/components/content-tab-assistant.tsx#L216-L225).

Commit:
Remove local message

Commit:
Attempt to remove local message

sejas reviewed (commented):
@katinthehatsite , As discussed in huddle, let's refactor this PR and create two smaller PRs.
- One to display the failed messages with a text notice.
- The second to add the button to retry the last message.

katinthehatsite commented:
Thanks @sejas - sounds good! I will proceed with this approach 👍

sejas commented:
Kat, I'm closing this PR because we already have each part in place:
- https://github.com/Automattic/studio/pull/315
- https://github.com/Automattic/studio/pull/319
